### PR TITLE
feat(helm chart): updated openebs helm chart to support multiple storage engine installation

### PR DESCRIPTION
**Pull request by shovanmaity**
Some components and csi-drivers are added as a dependency. This will enable us to install multiple storage engines using this umbrella chart.

openebs-ndm -> https://openebs.github.io/node-disk-manager
localpv-provisioner -> https://openebs.github.io/dynamic-localpv-provisioner
cstor -> https://openebs.github.io/cstor-operators
jiva -> https://openebs.github.io/jiva-operator
zfs-localpv -> https://openebs.github.io/zfs-localpv

Developer doc -> https://docs.google.com/document/d/1RE3rKWnsn3rmzPJ-ZtVCe6NvIH8o_FT2VKoh8UbSYBw
Storage engine specific configuration present as comment.
Migration details present in comment.

## Common upgrade path

```bash
# Install older openebs version using helm chart
helm install openebs \
--namespace openebs \
--create-namespace \
--version=2.8.0
# Upgrade to latest 2.X release
# a. delete some component to avoid error on (immutable match selector)
helm upgrade openebs \
--namespace openebs \
--set ndm.enabled=false \
--set ndmOperator.enabled=false \
--set localprovisioner.enabled=false \
--version=2.9.1
# b. earlier some storage classes are created by old modules now they are moved to
# helm chart so we need to import those resources. To do this you need to add some
# label and annotation on those storageclass(openebs-device,openebs-hostpath).
kubectl label sc <name> app.kubernetes.io/managed-by=Helm
kubectl annotate sc <name> meta.helm.sh/release-name=<release name>
kubectl annotate sc <name> meta.helm.sh/release-namespace=<release namespace>
# c. install ndm and local pv from dependency chart and install cstor and jiva csi driver
# if you have cstor or jiva volume. As this will be last release of 2.X and 3.X everything
# will be migrated to csi driver, you have to migrate non csi volumes to to csi volumes.
# For migration we need both csi and non csi components.
helm upgrade openebs \
--namespace openebs \
--set ndm.enabled=false \
--set ndmOperator.enabled=false \
--set localprovisioner.enabled=false \
--set jiva.enabled=true \
--set cstor.enabled=true \
--set openebs-ndm.enabled=true \
--set localpv-provisioner.enabled=true \
--version=2.9.1
```

NOTE - If you have not change any values during installation then set command will work. If you changed some stuff then check the comments in values.yaml

### CODE REVIEW

**Comment by shovanmaity** [comment]
Q - I want to install only localpv how can I do that?
A - Local pv has a dependency on openebs-ndm. In this umbrella chart there are no of components which has openebs-ndm as dependency. Make sure openebs-ndm is disabled inside local pv and it is enabled in root level. Here is an example of  values.yaml
```yaml
# This is older template make sure it is disabled
localprovisioner:
  enabled: false
# This is older template make sure it is disabled
ndm:
  enabled: false
# This is older template make sure it is disabled
ndmOperator:
  enabled: false
# Make sure cstor storage engine is disabled
cstor:
  enabled: true
# Make sure jiva storage engine is disabled
jiva:
  enabled: true
# Make sure zfs localpv storage engine is disabled
zfs-localpv:
  enabled: false
# Make sure localpv storage engine is enabled
localpv-provisioner:
  enabled: true
  # local pv settings goes here
  # https://openebs.github.io/dynamic-localpv-provisioner
  # Make sure ndm is enabled in root level only
  openebsNDM:
    enabled: false
openebs-ndm:
  enabled: true
  # ndm settings goes here
  # https://openebs.github.io/node-disk-manager
```
Installation steps
```bash
helm repo add openebs https://openebs.github.io/charts
helm repo update
helm install openebs --namespace openebs openebs/openebs --create-namespace
```

**Comment by shovanmaity** [comment]
Q - I want to install only jiva how can I do that?
A - Jiva has a dependency on local pv and local pv has a dependency on openebs-ndm. In this umbrella chart there are no of components which has openebs-ndm as dependency. Make sure local pv and openebs-ndm are disabled inside jiva and openebs-ndm is disabled inside local pv. Dependency local pv and openebs-ndm should be enabled at root level. Here is an example of  values.yaml
```yaml
# This is older template make sure it is disabled
localprovisioner:
  enabled: false
# This is older template make sure it is disabled
ndm:
  enabled: false
# This is older template make sure it is disabled
ndmOperator:
  enabled: false
# Make sure cstor storage engine is disabled
cstor:
  enabled: true
# Make sure jiva storage engine is enabled
jiva:
  enabled: true
  # jiva settings goes here
  # https://openebs.github.io/jiva-operator
# Make sure zfs localpv storage engine is disabled
zfs-localpv:
  enabled: false
# Make sure localpv storage engine is enabled
localpv-provisioner:
  enabled: true
  # local pv settings goes here
  # https://openebs.github.io/dynamic-localpv-provisioner
  # Make sure ndm is enabled in root level only
  openebsNDM:
    enabled: false
openebs-ndm:
  enabled: true
  # ndm settings goes here
  # https://openebs.github.io/node-disk-manager
```
Installation steps
```bash
helm repo add openebs https://openebs.github.io/charts
helm repo update
helm install openebs --namespace openebs openebs/openebs --create-namespace
```

**Comment by shovanmaity** [comment]
Q - I want to install only cstor how can I do that?
A - CStor has a dependency on openebs-ndm. In this umbrella chart there are no of components which has openebs-ndm as dependency. Make sure that openebs-ndm are disabled inside cstor. Dependency openebs-ndm should be enabled at root level. Here is an example of  values.yaml
```yaml
# This is older template make sure it is disabled
localprovisioner:
  enabled: false
# This is older template make sure it is disabled
ndm:
  enabled: false
# This is older template make sure it is disabled
ndmOperator:
  enabled: false
# Make sure cstor storage engine is enabled
cstor:
  enabled: true
  # cstor settings goes here
  # https://openebs.github.io/cstor-operators
# Make sure jiva storage engine is disabled
jiva:
  enabled: false
# Make sure zfs localpv storage engine is disabled
zfs-localpv:
  enabled: false
# Make sure localpv storage engine is disabled
localpv-provisioner:
  enabled: false
  openebsNDM:
    enabled: false
openebs-ndm:
  enabled: true
  # ndm settings goes here
  # https://openebs.github.io/node-disk-manager
```
Installation steps
```bash
helm repo add openebs https://openebs.github.io/charts
helm repo update
helm install openebs --namespace openebs openebs/openebs --create-namespace
```

**Comment by shovanmaity** [comment]
Q - I want to install only zfs localpv how can I do that?
A - ZFS localpv does not have dependency. Here is an example of  values.yaml
```yaml
# This is older template make sure it is disabled
localprovisioner:
  enabled: false
# This is older template make sure it is disabled
ndm:
  enabled: false
# This is older template make sure it is disabled
ndmOperator:
  enabled: false
# Make sure cstor storage engine is enabled
cstor:
  enabled: false
  openebsNDM:
    enabled: false
# Make sure jiva storage engine is disabled
jiva:
  enabled: false
  openebsLocalpv:
    enabled: false
  localpv-provisioner:
    openebsNDM:
      enabled: false
# Make sure zfs localpv storage engine is enabled
zfs-localpv:
  enabled: true
  # zfs-localpv settings goes here
  # https://openebs.github.io/zfs-localpv
# Make sure localpv storage engine is disabled
localpv-provisioner:
  enabled: false
  openebsNDM:
    enabled: false
openebs-ndm:
  enabled: false
```
Installation steps
```bash
helm repo add openebs https://openebs.github.io/charts
helm repo update
helm install openebs --namespace openebs openebs/openebs --create-namespace
```

**Comment by shovanmaity** [comment]
Q - What is the dependency tree and how can I install multiple storage engine
A - Here is the dependency tree of each individual component. Don't enable any sub dependency or dependency of any sub dependency, add them as a dependency in this chart and enable that at root level

- node disk manager
- dynamic localpv provisioner
  - node disk manager
- jiva operator
  - dynamic localpv provisioner
    - node disk manager
- cstor operators
  - node disk manager
- zfs localpv

**Comment by shovanmaity** [comment]
Q - I have installed jiva using [jiva-operator](https://openebs.github.io/jiva-operator) how can I migrate to [openebs](https://openebs.github.io/charts) chart?
```
# local pv is installed like this
helm repo add openebs-jiva https://openebs.github.io/jiva-operator
helm install openebs openebs-jiva/jiva -f values-jiva.yaml

# Now grab the configuration for jiva for openebs chart and do helm upgrade of same release
helm repo add openebs https://openebs.github.io/charts
helm install openebs openebs/openebs -f values-openebs.yaml
```

**Comment by shovanmaity** [comment]
Q - I have installed cstor using [cstor-operators](https://openebs.github.io/cstor-operators) how can I migrate to [openebs](https://openebs.github.io/charts) chart?
```
# cstor was installed like this
helm repo add openebs-cstor https://openebs.github.io/cstor-operators
helm install openebs openebs-cstor/cstor -f values-cstor.yaml

# Now grab the configuration for cstor for openebs chart and do helm upgrade of same release
helm repo add openebs https://openebs.github.io/charts
helm install openebs openebs/openebs -f values-openebs.yaml
```

**Comment by shovanmaity** [comment]
Q - I installed zfs localpv using [zfs-localpv](https://openebs.github.io/zfs-localpv) how can I migrate to [openebs](https://openebs.github.io/charts) chart?
```
# zfs localpv was installed like this
helm repo add openebs-zfslocalpv https://openebs.github.io/zfs-localpv
helm install openebs openebs-zfslocalpv/zfs-localpv -f values-zfs-localpv.yaml

# Now grab the configuration for zfx localpv for openebs chart and do helm upgrade of same release
helm repo add openebs https://openebs.github.io/charts
helm install openebs openebs/openebs -f values-openebs.yaml
```

**Comment by shovanmaity** [comment]
Q - How can I migrate to [openebs](https://openebs.github.io/charts) chart 2.x to 3.x?

1st you need to uninstall ndm and local pv component using this helm configuration. Still openebs version will be same
```yaml
localprovisioner:
  enabled: false
ndm:
  enabled: false
ndmOperator:
  enabled: false
```
Then you need to grab the the updated configuration for 3.x then you can perform helm upgrade and openebs version will be upgraded. One sample configuration is here -
```yaml
openebs-ndm:
  enabled: false
localpv-provisioner:
  enabled: false
  openebsNDM:
    enabled: false
```

**Review by sonasingh46 — APPROVED** [review]
lgtm
One question : 
- If let us say, standalone cStor is deployed using the cStor charts. And if an upgrade happens, using this all in one openebs chart the NDM will also get upgraded, right?

File: charts/openebs/values.yaml [hunk 1]
```diff
@@ -220,30 +230,16 @@ webhook:
     #   cpu: 250m
     #   memory: 500Mi
 
-jiva:
-  image: "openebs/jiva"
-  imageTag: "2.9.0"
-  replicas: 3
-  defaultStoragePath: "/var/openebs"
```
Comment (shovanmaity) [inline charts/openebs/values.yaml:227]: @shubham14bajpai in jiva csi how we are passing this value?
Is this a local pv config in jiva?

**Comment by shovanmaity** [comment]
> lgtm
> One question :
> 
> * If let us say, standalone cStor is deployed using the cStor charts. And if an upgrade happens, using this all in one openebs chart the NDM will also get upgraded, right?

Yes

File: charts/openebs/values.yaml (hunk 1 above)
Comment (shubham14bajpai) [reply charts/openebs/values.yaml:227]: This is not required in jiva csi as there is a policy which takes openebs-hostpath as replica storageclass

**Review by prateekpandey14 — COMMENTED** [review]
looks good, can we update the README with example installation helm chart commands with engine flag commands

**Review by akhilerm — APPROVED** [review]
LGTM

**Comment by shovanmaity** [comment]
> looks good, can we update the README with example installation helm chart commands with engine flag commands

:+1:  Do I need to push a doc or readme in this pr or in a separate PR?
IMO we can push this PR then if looks good then I will test the changes with hosted helm chart and make the doc. I am also ok with push it in the same PR.

**Review by kmova — APPROVED** [review]

**Comment by sonasingh46** [comment]
There was one little thing re docs -- if docs come in next PR, helm chart needs to be released again. Had a discussion with @prateek14 on this.
Not a big deal though
My bad I just missed it and should have put it earlier. I saw your approval notification and added a hold-merge anticipating a coming merge.

**Pull request merged**